marimo/_save/save.py

```diff
@@ -0,0 +1,220 @@
+# catch all to motive users to report if something does fail.
+UNEXPECTED_FAILURE_BOILERPLATE = (
+    "— this is"
+    " unexpected and is likely a bug in marimo. "
+    "Please file an issue at "
+    "https://github.com/marimo-team/marimo/issues"
+)
+
+
+if TYPE_CHECKING:
+    from types import FrameType, TracebackType
+
+    from _typeshed import TraceFunction
+    from typing_extensions import Self
+
+
+class SkipWithBlock(Exception):
+    """Special exception to get around executing the with block body."""
+
+
+class CacheException(BaseException):
+    pass
+
+
+class persistent_cache(object):
+    """Context block for cache lookup of a block of code.
+
+    Example usage:
+
+    >>> with persistent_cache(name="my_cache"):
+    >>>     variable = expensive_function() # This will be cached.
+
+    For an implementation sibling regarding the block skipping, see `withhacks`
+    in pypi.
+
+    NB: Since context abuses sys frame trace, this may conflict with debugging
+    tools that also use sys.settrace.
+    """
+
+    def __init__(
+        self,
+        *,
+        save_path: str = "outputs",
+        name: str,
+        _loader: Optional[Loader] = None,
+    ) -> None:
+        self.name = name
+        if _loader:
+            self._loader = _loader
+        else:
+            self._loader = PickleLoader(name, save_path)
+
+        self._skipped = True
+        self._cache: Optional[Cache] = None
+        self._entered_trace = False
+        self._old_trace: Optional[TraceFunction] = None
+        self._frame: Optional[FrameType] = None
+
+    def __enter__(self) -> Self:
+        sys.settrace(lambda *_args, **_keys: None)
+        frame = sys._getframe(1)
+        # Attempt to hold on to the previous trace.
+        self._old_trace = frame.f_trace
+        # Setting the frametrace, will cause the function to be run on _every_
+        # single context call until the trace is cleared.
+        frame.f_trace = self.trace
+        return self
+
+    def trace(
+        self, with_frame: FrameType, _event: str, _arg: Any
+    ) -> Union[TraceFunction | None]:
+        # General flow is as follows:
+        #   1) Follow the stack trace backwards to the first instance of a
+        # "<module>" function call, which corresponds to a cell level block.
+        #   2) Run static analysis to determine whether the call meets our
+        # criteria. The procedure is a little brittle as such, certain contexts
+        # are not allow (e.g. called within a function or a loop).
+        #  3) Hash the execution and lookup the cache, and return!
+        #  otherwise) Set _skipped such that the block continues to execute.
+
+        self._entered_trace = True
+        stack = traceback.extract_stack()
+
+        if not self._skipped:
+            return self._old_trace
+
+        # This is possible if `With` spans multiple lines.
+        if self._cache and self._cache.hit:
+            raise SkipWithBlock()
```

Comment:
I don't follow, how is a cache hit related to `With` spanning multiple blocks?

Reply:
Very weird, I think it has to do with compiling the module prior to traversing it. A direct attempt to replicate this bug does not work: https://marimo.app/l/bemoyu

But it definitely exists by this failing test: https://github.com/marimo-team/marimo/actions/runs/9947523119/job/27480334103?pr=1758